text/skia_text_blob_shaper.cpp

```diff
@@ -131,17 +132,33 @@ TextBlobRef SkiaTextBlob::MakeWithShaper(
   ASSERT(dynamic_cast<SkiaFont*>(font.get()));
 
   SkFont skFont = static_cast<SkiaFont*>(font.get())->skFont();
+  auto skFontMgr = static_cast<SkiaFontMgr*>(fontMgr.get())->skFontMgr();
   sk_sp<SkTextBlob> textBlob;
   gfx::RectF bounds;
-  if (auto shaper = SkShaper::Make(
-        static_cast<SkiaFontMgr*>(fontMgr.get())->skFontMgr())) {
+  if (auto shaper = SkShaper::Make(skFontMgr)) {
     ShaperRunHandler shaperHandler(text.c_str(), { 0, 0 }, handler);
-    shaper->shape(
-      text.c_str(), text.size(),
-      skFont,
-      true,
-      std::numeric_limits<float>::max(),
-      &shaperHandler);
+
+    auto bidiRun = SkShaper::MakeBiDiRunIterator(text.c_str(), text.size(), 0xfe);
+    constexpr SkFourByteTag tag = SkSetFourByteTag('Z', 'y', 'y', 'y');
+    auto scriptRun = SkShaper::MakeScriptRunIterator(text.c_str(), text.size(), tag);
+    auto languageRun = SkShaper::MakeStdLanguageRunIterator(text.c_str(), text.size());
+    auto fontRun = SkShaper::MakeFontMgrRunIterator(text.c_str(),
+                                                    text.size(),
+                                                    skFont,
+                                                    skFontMgr,
+                                                    "Arial", // Fallback
+                                                    SkFontStyle::Normal(),
+                                                    &*languageRun);
+
+    shaper->shape(text.c_str(),
+                  text.size(),
+                  *fontRun,
+                  *bidiRun,
+                  *scriptRun,
+                  *languageRun,
+                  std::numeric_limits<float>::max(),
+                  &shaperHandler);
+
```

Comment:
It looks like `MakeBiDiRunIterator` and `MakeScriptRunIterator` functions are legacy too.

Probably we could do something similar to `SkShaperPrimitive::shape()` (using values instead of smart pointers for each run iterator):
```c++
void SkShaperPrimitive::shape(const char* utf8,
                              size_t utf8Bytes,
                              const SkFont& font,
                              bool leftToRight,
                              SkScalar width,
                              RunHandler* handler) const {
    std::unique_ptr<FontRunIterator> fontRuns(
            MakeFontMgrRunIterator(utf8, utf8Bytes, font, nullptr));
    if (!fontRuns) {
        return;
    }
    // bidi, script, and lang are all unused so we can construct them with empty data.
    TrivialBiDiRunIterator bidi{0, 0};
    TrivialScriptRunIterator script{0, 0};
    TrivialLanguageRunIterator lang{nullptr, 0};
    return this->shape(utf8, utf8Bytes, *fontRuns, bidi, script, lang, nullptr, 0, width, handler);
}
```
Anyway I tried to replace with these iterators and didn't work (I got a crash), so there is something I'm missing here.

Reply:
I'll look into it, it's very strange that we're getting different behaviors per-platform

Reply:
Oh in this case your code worked on Linux :+1: I was just commenting this because it might be better if we avoid calling those legacy `Make*RunIterator` functions and create those iterators in the stack directly.

Reply:
The thing is that we need to use these functions because they have different implementations depending on what's available, so on a platform without harfbuzz or on a compile without it, it'll use a trivial one - etc. If we don't use them we'd have to include that logic ourselves.

The legacy stuff is kind of interesting too because the only place where that `SK_DISABLE_LEGACY_SKSHAPER_FUNCTIONS` is declared is in the Bazel compile scripts, not in the GN ones, so it's not even possible to disable these functions for anyone building aseprite atm. I think it's probably ok to keep using them for now.